dev/tools/gen_localizations.dart

```diff
@@ -104,132 +121,232 @@ String generateTranslationBundles() {
       }
-     output.writeln('''
-}''');
+     output.writeln('}');
+    }
+    if (count == 0) {
+      supportedLocales.writeln('///  * `$languageName` - ${describeLocale(languageName)}');
+    } else if (count == 1) {
+      supportedLocales.writeln('///  * `$languageName` - ${describeLocale(languageName)} (plus one variant)');
+    } else {
+      supportedLocales.writeln('///  * `$languageName` - ${describeLocale(languageName)} (plus $count variants)');
     }
   }
 
-  // Generate the translationBundleForLocale function. Given a Locale
-  // it returns the corresponding const TranslationBundle.
+  // Generate the getTranslation function. Given a Locale it returns the
+  // corresponding const GlobalMaterialLocalizations.
   output.writeln('''
 
-TranslationBundle translationBundleForLocale(Locale locale) {
+/// The set of supported languages, as language code strings.
+///
+/// The [GlobalMaterialLocalizations.delegate] can generate localizations for
+/// any [Locale] with a language code from this set, regardless of the region.
+/// Some regions have specific support (e.g. `de` covers all forms of German,
+/// but there is support for `de-CH` specifically to override some of the
+/// translations for Switzerland).
+///
+/// See also:
+///
+///  * [getTranslation], whose documentation describes these values.
+final Set<String> kSupportedLanguages = new HashSet<String>.from(const <String>[
+${languageCodes.map((String value) => "  '$value', // ${describeLocale(value)}").toList().join('\n')}
+]);
+
+/// Creates a [GlobalMaterialLocalizations] instance for the given `locale`.
+///
+/// The other arguments must give the arguments that will be passed to the [new
+/// GlobalMaterialLocalizations] constructor.
+///
+/// The following locales are supported by this package:
+///
+/// {@template flutter.localizations.languages}
+$supportedLocales/// {@endtemplate}
+///
+/// Generally speaking, this method is only intended to be used by
+/// [GlobalMaterialLocalizations.delegate].
+GlobalMaterialLocalizations getTranslation(
+  Locale locale,
+  intl.DateFormat fullYearFormat,
+  intl.DateFormat mediumDateFormat,
+  intl.DateFormat longDateFormat,
+  intl.DateFormat yearMonthFormat,
+  intl.NumberFormat decimalFormat,
+  intl.NumberFormat twoDigitZeroPaddedFormat,
+) {
   switch (locale.languageCode) {''');
+  const String arguments = 'fullYearFormat: fullYearFormat, mediumDateFormat: mediumDateFormat, longDateFormat: longDateFormat, yearMonthFormat: yearMonthFormat, decimalFormat: decimalFormat, twoDigitZeroPaddedFormat: twoDigitZeroPaddedFormat';
   for (String language in languageToLocales.keys) {
     if (languageToLocales[language].length == 1) {
       output.writeln('''
-    case \'$language\':
-      return const _Bundle_${languageToLocales[language][0]}();''');
+    case '$language':
+      return new MaterialLocalization${camelCase(languageToLocales[language][0])}($arguments);''');
```

Comment:
The arguments (date and time formats) are passed in. I may do a second pass that merges the gen_date_localization work into this too to avoid this problem, but I figured I should scope this work if I had any hope of getting anyone to review it.

Reply:
(I added a cache for these objects so that the benefits of `const`ness aren't lost.)